nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/security/RetentionExpiredAction.java

```diff
@@ -0,0 +1,132 @@
+                               Arrays.asList(INPUT_1 + ":" + ACTION_NAME, OUTPUT_1 + ":" + STATUS_STREAM))
+                       .build();
+    }
+
+    public static class RetentionExpiredComputation extends AbstractBulkComputation {
+
+        private static final Logger log = LogManager.getLogger(RetentionExpiredComputation.class);
+
+        public RetentionExpiredComputation() {
+            super(ACTION_NAME);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+            Collection<DocumentRef> refs = ids.stream().map(IdRef::new).collect(Collectors.toList());
+            for (DocumentRef ref : refs) {
+                // sanity checks
+                if (!session.isRecord(ref)) {
+                    log.debug("Document {} is not a record", ref);
+                    continue;
+                }
+                Calendar retainUntil = session.getRetainUntil(ref);
+                if (retainUntil == null) {
+                    log.debug("Document {} was not under retention", ref);
+                    continue;
+                }
+                if (isUnderRetention(retainUntil)) {
+                    log.debug("Document {} is still under retention: {}", ref, retainUntil.toInstant());
+                    continue;
+                }
+
+                try {
+                    // reset retention to null
+                    session.setRetainUntil(ref, null);
+                    // send event about retention expired
+                    sendEvent(session, ref, retainUntil);
+                } catch (NuxeoException e) {
+                    // TODO send to error stream
+                    log.warn("Cannot reset retention on: {}", ref, e);
+                }
+
+            }
+            try {
+                session.save();
+            } catch (PropertyException e) {
+                // TODO send to error stream
+                log.warn("Cannot save session", e);
+            }
+        }
+
+        protected static boolean isUnderRetention(Calendar retainUntil) {
+            return retainUntil != null && Calendar.getInstance().before(retainUntil);
+        }
+
+        protected void sendEvent(CoreSession session, DocumentRef ref, Calendar retainUntil) {
+            DocumentModel doc = session.getDocument(ref);
+            DocumentEventContext ctx = new DocumentEventContext(session, null, doc);
+            ctx.setProperty(CoreEventConstants.REPOSITORY_NAME, session.getRepositoryName());
+            ctx.setProperty(CoreEventConstants.SESSION_ID, session.getSessionId());
+            ctx.setProperty("category", DocumentEventCategories.EVENT_DOCUMENT_CATEGORY);
+            ctx.setProperty(CoreEventConstants.RETAIN_UNTIL, retainUntil);
+            ctx.setProperty("comment", retainUntil.toInstant().toString());
+            Event event = ctx.newEvent(DocumentEventTypes.RETENTION_EXPIRED);
```

Comment:
Does this trigger an elastic reindex?

Reply:
Indirectly yes, now, through the `setRetainUntil` above that resets retention to `null` and sends an event